In [ ]:
# 1. Cargar un archivo CSV desde el equipo
# IMPORTACIONES
import pandas as pd
from google.colab import files

# 1. Cargar un archivo CSV
files.upload()  # Subir archivo 'ejemplo_datos.csv'
df = pd.read_csv("ejemplo_datos.csv")
print("Archivo CSV básico:")
print(df.head())


FileNotFoundError: [Errno 2] No such file or directory: 'ejemplo_datos.csv'

In [ ]:
# 2. Cargar archivo Excel
# pip install openpyxl
files.upload()  # Subir archivo 'ejemplo_datos.xlsx'
df_excel = pd.read_excel("ejemplo_datos.xlsx", sheet_name=0)
print("\nArchivo Excel básico:")
print(df_excel.head())

# ¿Y si tu archivo tiene otra hoja?
# Puedes cambiar sheet_name=0 por:
# El nombre de la hoja:
# pd.read_excel("ejemplo_datos.xlsx", sheet_name="Hoja1")
# O también listar los nombres disponibles:
xls = pd.ExcelFile("ejemplo_datos.xlsx")
print(xls.sheet_names)


FileNotFoundError: [Errno 2] No such file or directory: 'ejemplo_datos.xlsx'

Leer un archivo Excel con nulos personalizados significa que, al cargar un archivo .xlsx en Pandas, puedes indicarle explícitamente qué valores deben considerarse como datos faltantes (NaN), aunque no estén escritos como tal en el archivo. Esto es útil porque en muchas hojas de Excel, los datos ausentes o inválidos no se marcan como NaN, sino con:

"NA", "N/A", "s/d", "-", "" (celda vacía), etc.

In [ ]:
# 3. Personalizar lectura CSV (separador y valores nulos)
files.upload()  # Subir archivo 'ejemplo_datos_personalizado.csv'
df_sep = pd.read_csv("ejemplo_datos_personalizado.csv", sep=';', na_values=['NA', 'N/A', ''])
print("\nCSV con separador ';' y nulos personalizados:")
print(df_sep.head())


Saving ejemplo_datos_personalizado.csv to ejemplo_datos_personalizado.csv

CSV con separador ';' y nulos personalizados:
   Nombre  Edad      Ciudad  Ventas
0     Ana  25.0    Santiago   100.0
1    Luis  30.0  Valparaíso     NaN
2  Carlos  35.0  Concepción   200.0
3   Marta   NaN    Rancagua   150.0
4   Pedro  28.0    Santiago     0.0


In [ ]:
# 4. Leer archivo Excel con nulos personalizados
files.upload()  # Subir archivo 'ejemplo_datos_personalizado.xlsx'
df_excel = pd.read_excel("ejemplo_datos_personalizado.xlsx", na_values=["NA", "N/A", ""])
print("\nExcel con valores nulos personalizados:")
print(df_excel.head())


Saving ejemplo_datos_personalizado.xlsx to ejemplo_datos_personalizado.xlsx

Excel con valores nulos personalizados:
   Nombre  Edad      Ciudad  Ventas
0     Ana  25.0    Santiago   500.0
1    Luis   NaN  Valparaíso     0.0
2  Carlos  30.0         NaN   750.0
3   Marta  35.0    Santiago   620.0
4   Pedro   0.0         NaN     NaN


In [ ]:
# 5. Verificar la cantidad de datos y tipos de columnas
print("Resumen de df (ejemplo_datos.csv):")
print(df.info())

print("\nResumen de df_excel (ejemplo_datos.xlsx):")
print(df_excel.info())

print("\nResumen de df_sep (ejemplo_datos_personalizado.csv):")
print(df_sep.info())

Resumen de df (ejemplo_datos.csv):


NameError: name 'df' is not defined

In [ ]:
# 6. Estadísticas descriptivas
print("\nEstadísticas descriptivas df:")
print(df.describe())

print("\nEstadísticas descriptivas df_excel:")
print(df_excel.describe())

print("\nEstadísticas descriptivas df_sep:")
print(df_sep.describe())


Estadísticas descriptivas df:
            Edad       Ventas
count   5.000000     4.000000
mean   33.800000  3400.000000
std     5.674504   787.400787
min    28.000000  2500.000000
25%    29.000000  2875.000000
50%    34.000000  3450.000000
75%    36.000000  3975.000000
max    42.000000  4200.000000

Estadísticas descriptivas df_excel:
            Edad      Ventas
count   5.000000    5.000000
mean   26.000000  556.000000
std    15.572412  346.164701
min     0.000000    0.000000
25%    25.000000  500.000000
50%    30.000000  620.000000
75%    35.000000  750.000000
max    40.000000  910.000000

Estadísticas descriptivas df_sep:
            Edad      Ventas
count   4.000000    4.000000
mean   29.500000  112.500000
std     4.203173   85.391256
min    25.000000    0.000000
25%    27.250000   75.000000
50%    29.000000  125.000000
75%    31.250000  162.500000
max    35.000000  200.000000


In [ ]:
# 7. Limpiar y calcular: reemplazar ventas nulas o negativas por la media y agregar columna de comisión y utilidad
def limpiar_y_calcular(df):
    if "Ventas" in df.columns:
        ventas_validas = df["Ventas"][(df["Ventas"] > 0) & (~df["Ventas"].isna())]
        media = ventas_validas.mean()
        df["Ventas"] = df["Ventas"].apply(lambda x: media if pd.isna(x) or x <= 0 else x)
        df["Comisión"] = df["Ventas"] * 0.10
        df["Utilidad"] = df["Ventas"] - df.get("Costos", 0) - df.get("Costos_Operativos", 0)
    return df

df = limpiar_y_calcular(df)
df_excel = limpiar_y_calcular(df_excel)
df_sep = limpiar_y_calcular(df_sep)

In [ ]:
# 8. Mostrar algunos resultados
print("\nDatos procesados df:")
print(df.head())

print("\nDatos procesados df_excel:")
print(df_excel.head())

print("\nDatos procesados df_sep:")
print(df_sep.head())


Datos procesados df:
   Nombre  Edad       Ciudad  Ventas  Comisión  Utilidad
0     Ana    28     Santiago  2500.0     250.0    2500.0
1    Luis    34   Valparaíso  3000.0     300.0    3000.0
2  Carlos    29     Rancagua  3400.0     340.0    3400.0
3   Marta    42    La Serena  4200.0     420.0    4200.0
4   Sofía    36  Antofagasta  3900.0     390.0    3900.0

Datos procesados df_excel:
   Nombre  Edad      Ciudad  Ventas  Comisión  Utilidad
0     Ana  25.0    Santiago   500.0      50.0     500.0
1    Luis   NaN  Valparaíso   695.0      69.5     695.0
2  Carlos  30.0         NaN   750.0      75.0     750.0
3   Marta  35.0    Santiago   620.0      62.0     620.0
4   Pedro   0.0         NaN   695.0      69.5     695.0

Datos procesados df_sep:
   Nombre  Edad      Ciudad  Ventas  Comisión  Utilidad
0     Ana  25.0    Santiago   100.0      10.0     100.0
1    Luis  30.0  Valparaíso   150.0      15.0     150.0
2  Carlos  35.0  Concepción   200.0      20.0     200.0
3   Marta   NaN    Ran

In [ ]:
# Exportar a archivos CSV
df.to_csv("df_limpio.csv", index=False)
df_excel.to_csv("df_excel_limpio.csv", index=False)
df_sep.to_csv("df_sep_limpio.csv", index=False)

In [ ]:
# Instalar openpyxl si aún no está
# !pip install openpyxl
import openpyxl as pd

# Exportar a archivos Excel
df.to_excel("df_limpio.xlsx", index=False)
df_excel.to_excel("df_excel_limpio.xlsx", index=False)
df_sep.to_excel("df_sep_limpio.xlsx", index=False)

In [ ]:
# Descargamos los archivos
from google.colab import files

# Descargar archivos CSV
files.download("df_limpio.csv")
files.download("df_excel_limpio.csv")
files.download("df_sep_limpio.csv")

# Descargar archivos Excel
files.download("df_limpio.xlsx")
files.download("df_excel_limpio.xlsx")
files.download("df_sep_limpio.xlsx")

FileNotFoundError: Cannot find file: df_limpio.csv

In [ ]:
# Reemplazar valores nulos o inválidos en Edad con la media (si tiene sentido hacerlo)
import pandas as pd
if "Edad" in df_excel.columns:
    edad_valida = df_excel["Edad"][pd.to_numeric(df_excel["Edad"], errors="coerce") > 0]
    edad_valida = pd.to_numeric(edad_valida, errors="coerce")
    media_edad = edad_valida.mean()
    df_excel["Edad"] = df_excel["Edad"].apply(lambda x: media_edad if pd.isna(x) or x <= 0 else x)

# Rellenar valores nulos en Ciudad con 'Desconocida'
if "Ciudad" in df_excel.columns:
    df_excel["Ciudad"] = df_excel["Ciudad"].fillna("Desconocida")

In [ ]:
# Descargar archivos Excel
df_excel.to_excel("df_excel_limpio_corregido.xlsx", index=False)
files.download("df_excel_limpio_corregido.xlsx")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
print(df.isna().sum())

Nombre      0
Edad        0
Ciudad      0
Ventas      0
Comisión    0
Utilidad    0
dtype: int64


TAREA: Análisis de datos de productos en tienda
Archivos requeridos:
productos.csv

ventas_tienda.xlsx

Archivo productos.csv (separador ;)
ID;Nombre;Categoría;Precio
101;Laptop;Electrónica;750
102;Celular;Electrónica;500
103;Zapatos;Ropa;80
104;Camisa;Ropa;NA
105;Tablet;Electrónica;-
106;Cartera;Accesorios;

Instrucciones:
1. Importa las librerías necesarias (pandas).

2. Carga los archivos respetando:

Para el CSV: Usa sep=';' y considera como nulos: "NA", "-" y "".

Para el Excel: considera como nulos: "NA", "-" y "".

3. Muestra las primeras filas de ambos DataFrames.

4. Cuenta cuántos datos nulos hay por columna en ambos archivos.

5. Llena los valores nulos en "Precio" con la media de los precios válidos.

6. Llena los valores nulos en "Ciudad" con "Desconocido".

7. Crea una nueva columna llamada Ingreso Total en el archivo Excel calculando Unidades * Precio (une ambas tablas por ID).

8. Filtra las filas donde el ingreso total sea mayor a 3000 y guarda esa tabla como ventas_altas.xlsx.

9. Exporta la tabla de productos corregida a un nuevo archivo llamado productos_limpios.csv.

In [ ]:
# 1.
import pandas as pd

In [ ]:
# 2. Cargar archivos y mostrar primeras filas
from google.colab import files

# Subir CSV
uploaded = files.upload()  # productos.csv
df_productos = pd.read_csv("productos.csv", sep=';', na_values=["NA", "-", ""])
print("Primeras filas de productos.csv:")
print(df_productos.head())

TypeError: 'NoneType' object is not subscriptable

In [ ]:
# Subir Excel
uploaded = files.upload()  # ventas_tienda.xlsx
df_ventas = pd.read_excel("ventas_tienda.xlsx", na_values=["NA", "-", ""])
print("\nPrimeras filas de ventas_tienda.xlsx:")
print(df_ventas.head())

Saving ventas_tienda.xlsx to ventas_tienda.xlsx

Primeras filas de ventas_tienda.xlsx:
   ID_Producto       Fecha  Unidades      Ciudad
0          101  2024-01-10       5.0    Santiago
1          102  2024-01-15       8.0         NaN
2          103  2024-01-20       NaN  Valparaíso
3          104  2024-01-22       3.0  Concepción
4          105  2024-01-25       NaN         NaN


In [ ]:
# 3. Contar valores nulos
print("\nNulos en productos.csv:")
print(df_productos.isna().sum())

print("\nNulos en ventas_tienda.xlsx:")
print(df_ventas.isna().sum())


Nulos en productos.csv:
ID           0
Nombre       0
Categoría    0
Precio       3
dtype: int64

Nulos en ventas_tienda.xlsx:
ID_Producto    0
Fecha          0
Unidades       2
Ciudad         2
dtype: int64


In [ ]:
# 4. Rellenar nulos
# Rellenar nulos en Precio con la media
media_precio = df_productos["Precio"].mean()
df_productos["Precio"] = df_productos["Precio"].fillna(media_precio)

# Rellenar nulos en Ciudad con "Desconocido"
df_ventas["Ciudad"] = df_ventas["Ciudad"].fillna("Desconocido")

# Rellenar nulos en Unidades con 0
df_ventas["Unidades"] = df_ventas["Unidades"].fillna(0)

In [ ]:
# Paso 5: Unir ambas tablas por ID y calcular "Ingreso Total"
# Unir df_ventas con df_productos según el ID de producto
df_ventas_completo = pd.merge(df_ventas, df_productos, how='left', left_on="ID_Producto", right_on="ID")

# Calcular el Ingreso Total: Unidades * Precio
df_ventas_completo["Ingreso_Total"] = df_ventas_completo["Unidades"] * df_ventas_completo["Precio"]

In [ ]:
# Paso 6: Filtrar ingresos mayores a 3000 y guardar en Excel
# Filtrar ventas con ingreso total mayor a 3000
ventas_altas = df_ventas_completo[df_ventas_completo["Ingreso_Total"] > 3000]

# Exportar a archivo Excel
ventas_altas.to_excel("ventas_altas.xlsx", index=False)

In [ ]:
# Paso 7: Exportar tabla de productos corregida a CSV
# Exportar la tabla de productos corregida
df_productos.to_csv("productos_limpios.csv", index=False, sep=";")
# Ver la tabla en pantalla
print(df_productos.head())  # Muestra las primeras filas


    ID   Nombre    Categoría      Precio
0  101   Laptop  Electrónica  750.000000
1  102  Celular  Electrónica  500.000000
2  103  Zapatos         Ropa   80.000000
3  104   Camisa         Ropa  443.333333
4  105   Tablet  Electrónica  443.333333


In [ ]:
# Para descargar el archivo .csv en tu computadora desde Colab:
from google.colab import files
files.download("productos_limpios.csv")

# Guardar ventas con ingresos mayores a 3000
ventas_altas.to_excel("ventas_altas.xlsx", index=False)
from google.colab import files
files.download("ventas_altas.xlsx")

FileNotFoundError: Cannot find file: productos_limpios.csv

**Datos de páginas web**

In [ ]:
# Leer una tabla directamente desde una página web (Web Scraping básico)
import pandas as pd

# URL con tablas en HTML (ejemplo: países por población)
url = "https://es.wikipedia.org/wiki/Anexo:Pa%C3%ADses_por_poblaci%C3%B3n"

# Leer todas las tablas de la página (devuelve una lista)
tablas = pd.read_html(url)

# Ver cuántas tablas encontró
print("Tablas encontradas:", len(tablas))

# Mostrar la primera tabla
df_web = tablas[0]
print(df_web.head())

Tablas encontradas: 7
  N.º País (o territorio dependiente)  \
0   1                           India   
1   2                      China[12]​   
2   3                  Estados Unidos   
3   4                       Indonesia   
4   5                        Pakistán   

  Proyección exponencial de la población al 1/7/2025[7]​ Total mun- dial (%)  \
0                                      1 417 492 000                    1754   
1                                      1 407 934 000                    1742   
2                                        342 181 000                     423   
3                                        285 783 000                     354   
4                                        256 204 000                     317   

  Cambio medio anual (%)[8]​ Cambio absoluto anual promedio  \
0                        090                     12 692 000   
1                       -005                     -1 390 000   
2                        061                      2 084 000  

Explicación:
pd.read_html(url) busca automáticamente todas las etiquetas <table> en la web.

Devuelve una lista de DataFrames.

Puedes acceder a cada tabla con índices como tablas[0], tablas[1], etc.

Se estraen 7 tablas en formato HTML (la mayoría en Wikipedia están muy bien estructuradas).

Mostrar la primera tabla, que contiene información poblacional por país, como India, China y EE.UU.
 ¿Qué muestra la salida?
Cada fila corresponde a un país y las columnas contienen datos como:

Población proyectada al 1/7/2025

Porcentaje mundial que representa

Cambio poblacional anual (en % y número absoluto)

Censo más reciente o estimación oficial

Fecha de esa estimación

Fuente o tipo de dato (A, E, R, etc.). A: Estimación de agencia	Datos generados por una agencia internacional (ej. ONU, Banco Mundial, etc.)
E:	Estimación nacional	Estimaciones producidas por el gobierno del país en base a modelos o proyecciones.

Enlace oficial de los datos





In [ ]:
# Ver nombre exacto de las columnas
print(df_web.columns)

Index(['N.º', 'País (o territorio dependiente)',
       'Proyección exponencial de la población al 1/7/2025[7]​',
       'Total mun- dial (%)', 'Cambio medio anual (%)[8]​',
       'Cambio absoluto anual promedio', 'Cambio medio abs. total anual (%)',
       'Años para even- tual du- pli- ca- ción[9]​',
       'Censo más reciente, última estimación oficial, proyección de la ONU o reloj de población nacional',
       'Fecha de esta última cifra de población en particular (d/mm/aaaa)[10]​',
       'Tipo[11]​',
       'Enlace o hipervínculo (usualmente de tipo oficial) de esta última cifra de población'],
      dtype='object')


In [ ]:
# Renombrar columnas para trabajar más fácilmente
df_web.columns = [
    "Ranking",
    "Pais",
    "Poblacion_2025",
    "Porcentaje_Mundial",
    "Cambio_Anual_Porc",
    "Cambio_Anual_Absoluto",
    "Cambio_Absoluto_Porc",
    "Anos_Duplicacion",
    "Censo_Ultimo",
    "Fecha_Censo",
    "Tipo",
    "Enlace"
]

In [ ]:
# Ver los primeros 5 países con mayor población proyectada
# Paso 1: Renombrar columnas para facilitar el trabajo
df_web.columns = [
    "Ranking",
    "Pais",
    "Poblacion_2025",
    "Porcentaje_Mundial",
    "Cambio_Anual_Porc",
    "Cambio_Anual_Absoluto",
    "Cambio_Absoluto_Porc",
    "Anos_Duplicacion",
    "Censo_Ultimo",
    "Fecha_Censo",
    "Tipo",
    "Enlace"
]

# Paso 2: Limpiar la columna de población (quitar espacios y caracteres invisibles)
df_web["Poblacion_2025"] = df_web["Poblacion_2025"].astype(str)                # Asegura tipo texto
df_web["Poblacion_2025"] = df_web["Poblacion_2025"].str.replace(r"[^\d]", "", regex=True)  # Deja solo números
#astype(str)
# Convierte todos los valores a cadena de texto.
# Esto es importante para luego aplicar métodos .str.
# .str es un acceso especial en Pandas que te permite aplicar métodos de cadenas (strings) a toda una columna de un DataFrame.
# Solo puedes usar .str en columnas de tipo texto (object o string).

# .str.replace(r"[^\d]", "", regex=True)
# Usa una expresión regular para eliminar todo lo que no sea un número (0–9).

# [^\d] = “todo carácter que no sea un dígito”

# "" = lo reemplaza con una cadena vacía

#Resultado: elimina espacios, puntos, comas, caracteres invisibles, etc.
# Ejemplo:
# Antes (texto original)	Después de limpieza
# "1 417 492 000"	            "1417492000"
# "342 181 000"	              "342181000"
# "–" (guion largo)         	"" (vacío)


# Paso 3: Convertir a número, errores a NaN
df_web["Poblacion_2025"] = pd.to_numeric(df_web["Poblacion_2025"], errors='coerce')
# Convierte la columna limpia a valores numéricos (int64 o float64).
# Si alguna celda no puede convertirse (por ejemplo, queda vacía o con caracteres extraños), se convierte en NaN (valor nulo).
# errors='coerce' = convierte errores a NaN en lugar de lanzar un error.

# Paso 4: Ordenar y mostrar los 5 primeros
df_top = df_web.sort_values("Poblacion_2025", ascending=False)
print(df_top[["Pais", "Poblacion_2025"]].head())
# Ordena el DataFrame df_web según la columna "Poblacion_2025".
# El parámetro ascending=False indica que el orden será descendente, es decir, de mayor a menor población.
# El resultado se guarda en df_top.

# Selecciona solo las columnas "Pais" y "Poblacion_2025" del DataFrame ordenado.
# Muestra las primeras 5 filas con .head() → que ahora son los 5 países con mayor población proyectada para 2025.


               Pais  Poblacion_2025
244           Mundo      8083785000
0             India      1417492000
1        China[12]​      1407934000
2    Estados Unidos       342181000
3         Indonesia       285783000


In [ ]:
# Contar cuántos países tienen datos censales reales
# Contar por tipo
conteo_tipos = df_web["Tipo"].value_counts()
print(conteo_tipos)
# df_web["Tipo"] accede a la columna "Tipo", que indica el tipo de fuente del dato poblacional.
# .value_counts() cuenta cuántas veces aparece cada tipo distinto.
# El resultado es una serie de Pandas con la frecuencia de cada tipo.

Tipo
E       97
C       37
O       25
A       24
M       17
F       15
T       11
R        9
P        9
B        1
Tipo     1
Name: count, dtype: int64


Cada letra en la columna "Tipo" representa el tipo de fuente o método usado para obtener la cifra de población. Aunque Wikipedia no siempre explica todos en detalle, estos suelen ser:

E: Estimación

C: Censo

O: Otra fuente oficial

A: Reloj de población automatizado

M: Modelado estadístico

F: Fuente oficial sin especificar tipo

T: Temporal o transitorio

R: Revisión o reproyección

P: Proyección futura

B: Base de datos u organismo especial

Tipo: Esta probablemente es una fila de encabezado mal interpretada como dato → puedes eliminarla.

In [ ]:
# Eliminar la fila errónea
df_web = df_web[df_web["Tipo"] != "Tipo"]

# Verifica nuevamente los tipos
conteo_tipos = df_web["Tipo"].value_counts()
print(conteo_tipos)

Tipo
E    97
C    37
O    25
A    24
M    17
F    15
T    11
R     9
P     9
B     1
Name: count, dtype: int64


TAREA GRUPAL Análisis de Datos de Productos en Tienda Objetivo: Aplicar conocimientos de limpieza y análisis de datos usando las librerías Pandas y NumPy en Python, trabajando con archivos CSV y Excel reales.

Modalidad: Trabajo en grupos de 5 o 4 estudiantes Archivos requeridos:
productos_grupal.csv
ventas_grupal.xlsx
Instrucciones:
Importar las librerías necesarias
import pandas as pd
Cargar los archivos respetando lo siguiente:
Para productos_grupal.csv: usar sep=";" y considerar como nulos los valores "NA", "-" y "".
Para ventas_grupal.xlsx: considerar los mismos valores nulos.
Mostrar las primeras 5 filas de ambos DataFrames.
Contar la cantidad de valores nulos por columna en ambos archivos.
Llenar los valores nulos en:
"Precio" con la media de precios válidos
"Ciudad" con "Desconocido"
"Unidades" con 0 si faltan (opcional: justificar si es adecuado)
Crear una nueva columna llamada Ingreso_Total en el DataFrame de ventas, calculando Unidades * Precio. Para ello, deben unir ambos DataFrames por la columna ID (o ID_Producto).
Filtrar las filas donde Ingreso_Total > 3000 y guardar ese nuevo DataFrame como ventas_mayores_3000.xlsx.
Exportar los datos limpios:
Guardar el DataFrame limpio de productos como productos_limpios.csv
Guardar el DataFrame de ventas limpio como ventas_limpias.xlsx
Responder en grupo:
¿Qué errores comunes encontraron en los datos?
¿Qué decisiones tomaron para tratarlos?
¿Qué ventaja tiene trabajar con Pandas frente a Excel en estos casos?
Entrega final:
Código en Google Colab o .ipynb exportado
Archivos generados: productos_limpios.csv y ventas_mayores_3000.xlsx




In [ ]:
df_productos.isnull().sum()
display(df_productos.isnull())

df_ventas.isnull().sum()
display(df_ventas.isnull())



NameError: name 'df_productos' is not defined